### PR TITLE
Create XRT_STATIC_COMPONENT option in cmake build files

### LINKED ISSUE #6799: Add CMAKE option to control building/installing static libraries
Currently, XRT has several libraries that it builds both shared and static and it seems in general to always explicitly specify either `SHARED` or `STATIC` when calling `add_library()` in cmake files.

For building XRT as a conda package, the community generally avoids providing static libraries for the reasons outlined in https://github.com/conda-forge/cfep/blob/main/cfep-18.md. 

Please add a cmake option that can be used to disable installing static libraries.  If XRT only cares to deliver STATIC or SHARED, you could switch to using[ the standard BUILD_SHARED_LIBS cmake flag](https://cmake.org/cmake/help/latest/variable/BUILD_SHARED_LIBS.html). But, if you want to keep delivering both, you will need to create your own XRT_INSTALL_STATIC_LIBS option.

### PR DESCRIPTION
**Pull request by timsnyder**
#### Problem solved by the commit
Want to separately package the static libraries.

#### ~~Bug~~ / issue (if any) fixed, which PR introduced the bug, how it was discovered
Fixes #6799 

#### How problem was solved, alternative solutions (if any) and why they were rejected
Alternate solution is to maintain this as a separate patch in the downstream package recipe and that was rejected because it will be too fragile to maintain.

#### Risks (if any) associated the changes in the commit
The only change that would be visible to those not setting `XRT_STATIC_COMPONENT` 

#### What has been tested and how, request additional testing if necessary
Hasn't been tested yet.  I'm planning to run the ctests in XRT against this with `XRT_STATIC_COMPONENT` at it's default value and to also package and test downstream against my design with manual spotchecking.  If there are more tests that I should run, I'd be happy to.

#### Documentation impact (if any)
If `XRT_DEV_COMPONENT` is documented somewhere, we might want to add `XRT_STATIC_COMPONENT` to wherever that might be.

### CODE REVIEW

**Comment by bryanloz-xilinx** [comment]
ok to test

**Comment by bryanloz-xilinx** [comment]
it was worth a try, but I guess I'm no admin :-)

File: src/CMakeLists.txt
Comment (bryanloz-xilinx) [inline src/CMakeLists.txt:93]: CMake options are used to set things ON or OFF, I could not get this working without converting the option to a cached CMake variable.  
```suggestion
set (XRT_DEV_COMPONENT "xrt" CACHE STRING "Set to 'xrt-dev' to split dev artifacts into separate component&package")
```

File: src/CMakeLists.txt
Comment (bryanloz-xilinx) [inline src/CMakeLists.txt:101]:
```suggestion
set (XRT_STATIC_COMPONENT "xrt" CACHE STRING "Set to 'xrt-static' to split static libs into separate component&package")
```

**Review by bryanloz-xilinx — CHANGES_REQUESTED** [review]

File: src/CMakeLists.txt
Comment (timsnyder) [inline src/CMakeLists.txt:93]: The logic CPack uses for it's configuration variables means that if you set `XRT_DEV_COMPONENT` to something other than the default `"xrt"` or `"xrt-dev"`, your `XRT_DEV_COMPONENT` package won't have the correct settings because cpackLin.cmake sets `CPACK_<PKG_TYPE>_XRT-DEV_*` variables and I didn't want to figure out it if it is possible to do something like:
`SET(CPACK_RPM_${XRT_DEV_COMPONENT}_PACKAGE_REQUIRES "<requirements>")`

I thought it was worth clarifying that you can only specify `"xrt-dev"`.

I also thought it was worth making `XRT_DEV_COMPONENT` a cached variable so that you could add it to the initial cmake options and the build would remember it for subsequent steps.  I can understand if you want to remove the caching and have this be a 'hidden' and unsupported feature of the build.

File: src/runtime_src/core/common/CMakeLists.txt
Comment (timsnyder) [inline src/runtime_src/core/common/CMakeLists.txt:97]: It seems a little weird to me that when `XRT_DEV_COMPONENT` is set to `xrt-dev`, you set the `xrt-dev` package to depend on `xrt` (the runtime package), but you also include all of the runtime package targets in what is being installed into `xrt-dev`.  It seems like you would have all the runtime files inside the `xrt-dev` package and not need the depenence on the runtime package.

If you were going to make the `xrt-dev` package depend on the `xrt` runtime package, I would think `xrt-dev` would only include the:
* headerfiles
* static libraries (unless you're splitting them out into `xrt-static`)
* cmake and pkg-config definition files

File: src/runtime_src/core/common/CMakeLists.txt
Comment (stsoe) [inline src/runtime_src/core/common/CMakeLists.txt:107]: I don't understand why we would need a static package / target?

File: src/runtime_src/xocl/CMakeLists.txt
Comment (timsnyder) [inline src/runtime_src/xocl/CMakeLists.txt:115]:
```suggestion
install(TARGETS xilinxopencl_static
```
oops.  this was a typo.

**Review by stsoe — CHANGES_REQUESTED** [review]
Please review my comments and educate me.  THanks!

File: build/build.sh
Comment (timsnyder) [reply build/build.sh:232]: Oh, I didn't think that `XRT_DEV_COMPONENT` was being used to deliver the released packages  and I was trying not to change the default behavior of your build.  

My reference for whether it was being used is that when I go to https://www.xilinx.com/products/boards-and-kits/alveo/u250.html#gettingStarted and select an OS&Version combination, I'm only asked to install one package for XRT that contains the headers and libraries called 'xrt'.  I will admit that my knowledge of XRT use-cases is very limited... :)

File: build/build.sh
Comment (timsnyder) [reply build/build.sh:63]: Some downstream packaging ecosystems split static libraries into their own packages as a matter of policy.  At a high-level, splitting them into a separate package enables the package ecosystem to clearly distinguish packages in their ecosystem depend on the static libraries and must be rebuilt to benefit from any change in the library versus packages that use dynamic linkage and will pick up the new library version (provided that their requirements allow installation of the new version).

For more detailed explaination specific to the [conda-forge](https://conda-forge.org/) ecosystem, see https://github.com/conda-forge/cfep/blob/main/cfep-18.md

File: build/build.sh
Comment (timsnyder) [reply build/build.sh:227]: The dynamic libraries are useful without the headers if you are only interested in deploying an already compiled host application.  Most of the system package managers in Linux distributions ship a runtime package that only contains executables and dynamic-libs and a split 'development' package that contains headers and other things needed to link against the libraries.

I agree that the static libraries are useless without the headers and that's why the `xrt-static` package would depend on the `XRT_DEV_COMPONENT` package that contains the headers.  The intent is to allow users who do not want to ever accidentally link the static version of the library to install only the dynamic libs.

File: src/runtime_src/core/common/CMakeLists.txt
Comment (timsnyder) [reply src/runtime_src/core/common/CMakeLists.txt:107]: I think I answered this in response to your first comment at https://github.com/Xilinx/XRT/pull/6830#discussion_r912500947

**Comment by timsnyder** [comment]
@stsoe, Thank you for your review!

**Comment by timsnyder** [comment]
Looks like the version of Centos used by AWS:
```
 ___ ___  ___   _     ___  _____   __    _   __  __ ___
| __| _ \/ __| /_\   |   \| __\ \ / /   /_\ |  \/  |_ _|
| _||  _/ (_ |/ _ \  | |) | _| \ V /   / _ \| |\/| || |
|_| |_|  \___/_/ \_\ |___/|___| \_/   /_/ \_\_|  |_|___|
AMI Version:        1.12.0
```
is no longer new enough to build XRT latest.  

I've run `xrtdeps.sh` to get packages updated and when I run `cd build && ./build.sh -split_dev -split_static` (on this PR branch) I run into a bunch of CMake errors saying:
```
CMake Error in /home/centos/XRT/tests/validate/aie_pl_test/CMakeLists.txt:
  Target "aie_pl.exe" requires the language dialect "CXX17" (with compiler
  extensions), but CMake does not know the compile flags to use to enable it.
```
 
 I'm spinning up a newer VM.

**Comment by timsnyder** [comment]
I'm reminded that the AMI version can mean many things... This is what I was running on:
```
-bash-4.2$ cat /etc/os-release
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

```

File: build/build.sh
Comment (stsoe) [reply build/build.sh:227]: > The dynamic libraries are useful without the headers if you are only interested in deploying an already compiled host application. Most of the system package managers in Linux distributions ship a runtime package that only contains executables and dynamic-libs and a split 'development' package that contains headers and other things needed to link against the libraries.

Yes, this matches what I mentioned.

> I agree that the static libraries are useless without the headers and that's why the `xrt-static` package would depend on the `XRT_DEV_COMPONENT` package that contains the headers. The intent is to allow users who do not want to ever accidentally link the static version of the library to install only the dynamic libs.

I am not convinced that we need a static package.  AFAIK the *-dev packages always contains link libraries whether static or dynamic and headers.  I don't know about conda, but I know that XRT does not package for different environments, and I would hate to see Conda specific changes to XRT.

File: build/build.sh
Comment (stsoe) [reply build/build.sh:232]: We never officially supported a development package, and I am happy to see you enable that.  My point is that I expect two packages only: a runtime package and a development package.  And while not supported by build.sh (before your changes) it was supported by the various makefiles through the xrt-targets and xrt-dev-targets.

The build.sh option you add should simply enable the development / runtime split, it should not create a 3rd package.

**Comment by bryanloz-xilinx** [comment]
@timsnyder in CentOS you might need to do
scl enable devtoolset-9 bash

File: build/build.sh
Comment (bryanloz-xilinx) [reply build/build.sh:227]: > I am not convinced that we need a static package. AFAIK the *-dev packages always contains link libraries whether static or dynamic and headers. I don't know about conda, but I know that XRT does not package for different environments, and I would hate to see Conda specific changes to XRT. The accidental linking against dynamic vs static, is a pilot error.

@stsoe thanks for looking at this!

At the end of the day, if you don’t want to support separation of dynamic and static libraries, It’s your call, and we will respect it.

I just want to state that there are established conventions that most projects use, where -DBUILD_SHARED_LIBS=ON controls whether static or dynamic libraries are built. Protobuf, glog, gflags, json-c all obey. Their users may not want to waste disk space on static libraries if they will never use them. It really is not a conda specific request. Perhaps we should see how they implement this feature, as it may be simpler than adding a new component.

Comment (timsnyder) [reply build/build.sh:227]: > At the end of the day, if you don’t want to support separation of dynamic and static libraries, It’s your call, and we will respect it.

completely agree.

**Comment by timsnyder** [comment]
> scl enable devtoolset-9 bash

yeah, that's right.  I forgot that it is documented in the build instructions but not automated by build.sh.

File: build/build.sh
Comment (stsoe) [reply build/build.sh:227]: Yeah, I just don't believe in the disk space issue for developers.  I am absolutely in agreement regarding making runtime package as small as possible, but I have never seen the static library vs dynamic link libraries split you are talking about.  It really would require a total of 4 package for this to work

1. runtime
2. headers (dev)
3. dynamic link (dev)
4. static link (dev)

For a project the size of XRT this is an overkill IMO and in fact I don't recall ever seeing this kind of split in any packages I install with apt.   

Can we please just enable a runtime package and a development package?  Enabling this is long over due and I really appreciate you bringing this back to the surface for discussion and action.

**Comment by bryanloz-xilinx** [comment]
> > scl enable devtoolset-9 bash
> 
> yeah, that's right. I forgot that it is documented in the build instructions but not automated by build.sh.

You make a good point, it could easily be automated

**Comment by timsnyder** [comment]
Okay.  After remembering the devtoolset-9 activation, on Centos7, using cmake3 installed by xrtdeps.sh (verison 3.17.5), I see:
```
----XRT_BUILD_INSTALL_DIR=/home/centos/XRT/build/Debug/opt/xilinx/xrt
-- Configuring done
CMake Error: INSTALL(EXPORT) given unknown export "xrt-dev-targets"
-- Generating done
CMake Generate step failed.  Build files cannot be regenerated correctly.
bash-4.2$ which cmake3
/usr/bin/cmake3
bash-4.2$ cmake3 --version
cmake3 version 3.17.5
```

And I think that is https://gitlab.kitware.com/cmake/cmake/-/issues/17925, resolved in cmake 3.19.  Upgrading cmake to 3.19 gets further.  

@stsoe at this point, it looks like we will not merge this PR but I'd still like to document any issues I encounter while trying this out and I can use this branch to try out the split packaging before rolling back the creation of XRT_STATIC_COMPONENT.

I think it's possible that XRT_STATIC_COMPONENT is causing xrt-dev-targets to be empty at a point and cause cmake to hit this edgecase but I thought it was worth documenting.

**Comment by stsoe** [comment]
@timsnyder Thanks for the effort.  I would very much like to have runtime and development component supported without separation of static libraries.  I don't remember what the existing hooks in XRT CMakeLists.txt files required in terms of CMake version, I only remember that modern CMake had better support, that didn't require two install() calls.

File: build/build.sh
Comment (timsnyder) [reply build/build.sh:227]: I think that the dynamic link libraries are needed in runtime, along with any executables.  

You're right about debian and ubuntu not having this split.  I was curious what I could find about this topic outside of the conda ecosystem, so I did a little digging, just for my own entertainment:
* Debian
  * Debian policy says put the static libs in the dev package: https://www.debian.org/doc/debian-policy/ch-sharedlibs.html#static-libraries
  * However, they also have a page that talks about the downsides of static linking and how to mitigate them: https://wiki.debian.org/StaticLinking
    * they have a tool that identifies and tracks statically linked binaries in their packages https://lintian.debian.org/tags/statically-linked-binary
    * they also keep source code around for licensing reasons.  Since everything in XRT that ends up in a static library is Apache 2, perhaps this isn't an issue.  I'm still unclear as to exactly what in XRT gets infected with GPL, probably only the kernel modules but I personally haven't reviewed it.  Also, perhaps the availability of the source code on github satisfies the GPL requirements.  Again, I'm neither a licensing expert nor a lawyer.
 
* OpenSUSE
  * doesn't explicitly allow or disallow them but [says it is best practice to avoid them](https://en.opensuse.org/openSUSE:Shared_library_packaging_policy#exception_4:~:text=Avoid%20packaging%20static%20libraries.%20You%20should%20use%20%2D%2Ddisable%2Dstatic%20configure%20option%20or%2C%20as%20a%20last%20resort%2C%20remove%20static%20libraries%20after%20make%20install.%20If%20in%20doubt%2C%20ask.)
* Fedora
  * https://docs.fedoraproject.org/en-US/packaging-guidelines/#packaging-static-libraries starts "Packages including libraries SHOULD exclude static libs as far as possible"
  * Requires shipped static libs to be identified with a `-static` package (but allow it to be a 'virtual' package) and any users of the static libraries must declare their dependence on the `-static` package.  I thought this was an interesting idea and scanned through https://github.com/conda-forge/cfep/pull/34 (adopting cfep-18) to see if it was discussed.  I don't see it there.  I suspect that allowing the virtual `-static` package would be difficult in practice to enforce without support in the tooling.  I think cfep-18 takes a KISS approach by requiring the static-libs to be in a separate package
  
Got tired of looking right now but it might be interesting to see if Cannonical or RedHat have specific statements regarding the packaging of static libs.
  
  
I will acknowledge that XRT probably has one of the best use-cases for wanting to support statically linked executables by it's end users.  I would think most people that bundle their host executable code into the xclbin will want it to be statically linked so that they don't have to worry about keeping track of all the dynamic libs for deploying an app, it's just all bundled up nicely in the xclbin.

**Comment by chvamshi-xilinx** [comment]
ok to test

File: build/build.sh
Comment (stsoe) [reply build/build.sh:227]: > I think that the dynamic link libraries are needed in runtime, along with any executables.

No, definitely not.   On Linux, the link libraries are softlinks installed by the development package.  The softlinks are named `foo.so` and point to a versioned library (foo.so -> foo.so.2 -> foo.so.2.1.0) installed by the runtime package.

Comment (timsnyder) [reply build/build.sh:227]: Oh, ok, I understand what you were saying now.  The unversioned symbolic links to the dynamic libraries definitely do need to only be installed in the dev package.  I was confused and thought you meant all dynamic library artifacts.

**Comment by stsoe** [comment]
> @stsoe, I suppose there is not an easy way for me to see the output of your build pipeline? The link seems to point to in internal host. I'd be curious to see where it is failing because it builds for me locally.

23:05:23 CMake Error: INSTALL(EXPORT) given unknown export "xrt-dev-targets"
Centos7.8, Ubuntu18/20.04, AmazonLinux

**Comment by timsnyder** [comment]
Ahh.  Thanks.  That's the CMake bug I mentioned in https://github.com/Xilinx/XRT/pull/6830#issuecomment-1173161053.  I had forgotten about that and it makes sense that you hit it in this build.  I'll make sure to check for it when I drop back to only haveing the runtime and dev packages.

**Review by rozumx — APPROVED** [review]
Syntax and changes looks good.

**Comment by stsoe** [comment]
Closing since we don't want separate static package.

**Pull request closed without merging**
